src/Raven.Server/Documents/Handlers/CounterHandler.cs

```diff
@@ -0,0 +1,237 @@
+﻿// -----------------------------------------------------------------------
+//  <copyright file="CounterHandler.cs" company="Hibernating Rhinos LTD">
+//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
+//  </copyright>
+// -----------------------------------------------------------------------
+
+using System.Net;
+using System.Threading.Tasks;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.Handlers
+{
+    public class CounterHandler : DatabaseRequestHandler
+    {
+        public class IncrementCounterCommand : TransactionOperationsMerger.MergedTransactionCommand
+        {
+            private DocumentDatabase _database;
+            private string _doc, _counter;
+            private long _value;
+
+            public long CurrentValue;
+
+            public IncrementCounterCommand(DocumentDatabase database, string doc, string counter, long value)
+            {
+                _database = database;
+                _doc = doc;
+                _counter = counter;
+                _value = value;
+            }
+
+            public override int Execute(DocumentsOperationContext context)
+            {
+                _database.DocumentsStorage.CountersStorage.IncrementCounter(context, _doc, _counter, _value);
```

Comment:
Here we need to validate that the document exists.
We'll add it to the flags as HasCounters here